cam6_3_001: Katec/ifsall3 changes

Katetc opened the pull request:
Fixes #162 PUMAS SCAM and IFS updates

. Add diagnostic output fields (3) to mg3.
. Add namelist switches to pumas to reproduce ECMWF IFS microphysics behavior
. Add a SCAM case for the S. Ocean 'micre2017'
. Fix a bug in the scam scripts created when usermods_dirs was moved

Because of changes in a few parameters used in simulations where the "constant number" flag is turned on (largely Aquaplanet compsets, and not any F or B configurations), this tag should be considered answer changing. The switches for IFS-like behavior do not need to be exposed to CESM namelists at this point. They will be set when needed (for high-end testing) via user mods. 

Also fixes #206 because Cheryl asked me to. :)

brian-eaton commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -298,7 +312,12 @@ subroutine micro_mg_cam_readnl(nlfile)
        micro_mg_do_hail, micro_mg_do_graupel,micro_mg_ngcons, micro_mg_ngnst,&
        micro_mg_nccons, micro_mg_nicons, micro_mg_ncnst, micro_mg_ninst,&
-       micro_do_massless_droplet_destroyer
+       micro_mg_nrcons, micro_mg_nscons, micro_mg_nrnst, micro_mg_nsnst,&
+       micro_do_massless_droplet_destroyer,&
+       micro_mg_evap_sed_off, micro_mg_icenuc_rh_off, micro_mg_icenuc_use_meyers, &
+       micro_mg_evap_scl_ifs, micro_mg_evap_rhthrsh_ifs, &
+       micro_mg_rainfreeze_ifs, micro_mg_ifs_sed, micro_mg_rain_fall_corr
```
I think the new switches should either be in the namelist (which requires namelist definition entries) or not in the namelist.  The current partial implementation is confusing.

gold2718 replied:
I agree. How is a user supposed to set values such as `micro_mg_ninst`? Being in the namelist, a user would likely try to put an entry in user_nl_cam which will blow up because there is no definition.

Katetc replied:
Hi Steve, did you see my comment below that I was just missing some changes in a namelist file? All of these options are now namelist definition entries.

Katetc replied:
Oh shoot. I tried to delete my own response here because I realized that Steve was concerned about the parameter values and not the logical switches. And it looks like maybe I deleted Steve's comment??? So sorry. To address parameter values, while we might want to expose those for a perturbed parameter ensemble or some other experiment, for the time being, those constant values are not to be changed by the user. The logical switches are all that should be exposed.

Katetc commented:
I addressed the suggestions here in my most recent commit. After chatting with Andrew, it turns out that I missed his modified namelist_definitions.xml file, so I added that in here and now all namelist variables ARE available in the namelist. I also fixed the comments and updated the scam scripts as Jessie suggested (that was a good idea, thanks!)

nusbaume commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2601,7 +2600,79 @@ Default: 0.1e6 m-3
+
+<entry id="micro_mg_nrnst" type="real" category="microphys"
+       group="micro_mg_nl" valid_values="" >
+In-cloud rain number concentration when micro_mg_nicons=.true.
```
I assume this is supposed to say:

`micro_mg_nrcons=.true.`

not

`micro_mg_nicons=.true.`

as it does currently?

nusbaume commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2601,7 +2600,79 @@ Default: 0.1e6 m-3
+
+<entry id="micro_mg_nsnst" type="real" category="microphys"
+       group="micro_mg_nl" valid_values="" >
+In-cloud snow number concentration when micro_mg_nicons=.true.
```
I assume this is supposed to say

`micro_mg_nscons=.true.`

here, and not `nicons`?

nusbaume commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -134,16 +134,30 @@ module micro_mg_cam
+real(r8) :: micro_mg_ncnst = 50.e6_r8 ! constant droplet num concentration (m-3)
+real(r8) :: micro_mg_ninst = 0.05e6_r8  ! constant ice num concentration (m-3)
+real(r8) :: micro_mg_nrnst = 0.2e6_r8     ! rain  num concentration when nrcons=.true. (m-3)
+real(r8) :: micro_mg_nsnst = 0.005e6_r8 ! ice num concentration when nicons=.true. (m-3)
```
It looks like this code comment issue is still here?  Of course if this is the correct comment then feel free to disregard this particular request.

nusbaume commented:
One other thing I forgot to mention, can you please replace `Fixes issue #162` with just `Fixes #162` in your top-level PR description?  The script that is used to automatically close issues currently needs the keyword and the issue number to be right next to each other in order to work properly (although I will hopefully loosen that restriction at some point in the future).  Thanks!

gold2718 commented on bld/scripts/create_scam6_iop_multi:
```diff
@@ -40,7 +40,15 @@ set COMPILER=intel
 
 ## location of IOP data in CESM Tag
-set MODSDIR = $CESMDIR/components/cam/cime_config/usermods_dirs
+## location of IOP data in CESM Tag                                                                    
```
Is this line really duplicated?
Also, OMG, we still have csh files?

gold2718 commented on cime_config/usermods_dirs/scam_micre2017/user_nl_cam:
```diff
@@ -0,0 +1,16 @@
+scmlon=$PTS_LON
+scmlat=$PTS_LAT
```
These are no longer supported (the model gets the correct values directly from the driver). Just remove these lines.

gold2718 reviewed (changes requested):
I think that @nusbaume's questions need to addressed along with my concerns.

Katetc commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2601,7 +2600,79 @@ Default: 0.1e6 m-3
        group="micro_mg_nl" valid_values="" >
 In-cloud graupel number concentration when micro_mg_nicons=.true.
-Default: 0.1e6 m-3
+Default: 0.0005e6 m-3
```
Easier to compare across values with the same exponent. You'll notice other "nst" values are 100e6 and 0.1e6.

Katetc commented:
Hi all, we added in one more change here: we changed the name of micro_mg_rain_fall_corr to micro_mg_precip_fall_corr because this logic is now applied to other falling species (snow and hail/graupel). This change is done in the new tag of PUMAS, which also includes changes from Adam Herrington (v1.6) that we rolled back (v1.7). If you are curious, you can see the changes in PUMAS using this comparison: https://github.com/ESCOMP/PUMAS/compare/pumas_cam-release_v1.5...pumas_cam-release_v1.8

andrewgettelman commented:
Whoops. Sorry to jump in here. 

There was one more change that does not seem to be in the code being reviewed. And I don't even know how to find that code. I apologize for jumping in to the discussion without the code mods in hand, but I'm not sure how to use the github workflow yet. 

Specifically, 

micro_mg_rain_fall_corr

should have been changed to micro_mg_precip_fall_corr 

in micro_mg3, micro_mg_cam and namelist_definitions

Since it is supposed to apply now to rain, snow and graupel. 

Sorry, but I want to flag this now and I can talk to Kate about it: I'm not even sure how I look at the code, and will try to find it.

nusbaume commented:
Hi Andrew,

You can find all of the code changes being reviewed by going to the
ESCOMP/CAM webpage here:

https://github.com/ESCOMP/CAM

Selecting the "Pull requests" tab near the top, selecting the "cam6_3_001"
pull request, and then clicking on the "Files changed" tab just below the
pull request title, which will show all of the code modifications that are
included in this particular PR.  This same method can be used to find code
changes for other PRs as well, with the only difference being the title of
that particular PR.

Also, it looks like Kate has already added the change you are requesting
(as I only see "micro_mg_precip_fall_corr"), so I believe this PR is ready
for code review, unless there are additional changes you would like to add
as well?

Anyways, I hope that helps, and have a great day!

Jesse



On Fri, Sep 18, 2020 at 9:31 AM Andrew Gettelman <notifications@github.com>
wrote:

> Whoops. Sorry to jump in here.
>
> There was one more change that does not seem to be in the code being
> reviewed. And I don't even know how to find that code. I apologize for
> jumping in to the discussion without the code mods in hand, but I'm not
> sure how to use the github workflow yet.
>
> Specifically,
>
> micro_mg_rain_fall_corr
>
> should have been changed to micro_mg_precip_fall_corr
>
> in micro_mg3, micro_mg_cam and namelist_definitions
>
> Since it is supposed to apply now to rain, snow and graupel.
>
> Sorry, but I want to flag this now and I can talk to Kate about it: I'm
> not even sure how I look at the code, and will try to find it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CAM/pull/178#issuecomment-694935987>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHLLQEJCF4DLVXR7DYVJ6STSGN4NXANCNFSM4PW2JKZA>
> .
>


-- 
Jesse Nusbaumer

Software Engineer II
Atmospheric Modeling and Predictability, CGD, NCAR
Phone #:   303-497-1381
Cell #:       260-602-2105
Github:      nusbaume
Pronouns: He/Him/His

andrewgettelman commented:
She did. Sorry, I thought I deleted the comment and I did find that. It was
buried in the thread that she updated it.

There was some link at the top of the page to the specific fork.

Again, my apologies for not understanding the process. Still hard for a
novice to follow.

A

On Fri, Sep 18, 2020 at 12:25 PM Jesse Nusbaumer <notifications@github.com>
wrote:

> Hi Andrew,
>
> You can find all of the code changes being reviewed by going to the
> ESCOMP/CAM webpage here:
>
> https://github.com/ESCOMP/CAM
>
> Selecting the "Pull requests" tab near the top, selecting the "cam6_3_001"
> pull request, and then clicking on the "Files changed" tab just below the
> pull request title, which will show all of the code modifications that are
> included in this particular PR. This same method can be used to find code
> changes for other PRs as well, with the only difference being the title of
> that particular PR.
>
> Also, it looks like Kate has already added the change you are requesting
> (as I only see "micro_mg_precip_fall_corr"), so I believe this PR is ready
> for code review, unless there are additional changes you would like to add
> as well?
>
> Anyways, I hope that helps, and have a great day!
>
> Jesse
>
>
>
> On Fri, Sep 18, 2020 at 9:31 AM Andrew Gettelman <notifications@github.com
> >
> wrote:
>
> > Whoops. Sorry to jump in here.
> >
> > There was one more change that does not seem to be in the code being
> > reviewed. And I don't even know how to find that code. I apologize for
> > jumping in to the discussion without the code mods in hand, but I'm not
> > sure how to use the github workflow yet.
> >
> > Specifically,
> >
> > micro_mg_rain_fall_corr
> >
> > should have been changed to micro_mg_precip_fall_corr
> >
> > in micro_mg3, micro_mg_cam and namelist_definitions
> >
> > Since it is supposed to apply now to rain, snow and graupel.
> >
> > Sorry, but I want to flag this now and I can talk to Kate about it: I'm
> > not even sure how I look at the code, and will try to find it.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/ESCOMP/CAM/pull/178#issuecomment-694935987>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AHLLQEJCF4DLVXR7DYVJ6STSGN4NXANCNFSM4PW2JKZA
> >
> > .
> >
>
>
> --
> Jesse Nusbaumer
>
> Software Engineer II
> Atmospheric Modeling and Predictability, CGD, NCAR
> Phone #: 303-497-1381
> Cell #: 260-602-2105
> Github: nusbaume
> Pronouns: He/Him/His
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESCOMP/CAM/pull/178#issuecomment-695018745>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACJL32WAMAVYPQ7PXHPSYILSGOQYZANCNFSM4PW2JKZA>
> .
>


-- 
Dr. Andrew Gettelman   andrew@ucar.edu
http://www.cgd.ucar.edu/staff/andrew/
+1.303.497.1887
National Center For Atmospheric Research
3090 Center Green Drive,
Boulder, Colorado 80301, USA

nusbaume commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -134,16 +134,30 @@ module micro_mg_cam
+logical  ::  micro_mg_evap_rhthrsh_ifs = .false.  ! Evap RH threhold following ifs
+logical  ::  micro_mg_rainfreeze_ifs = .false.    ! Rain freezing at 0C following ifs
+logical  ::  micro_mg_ifs_sed = .false.           ! Snow sedimentation = 1 m/s following ifs
+logical  ::  micro_mg_precip_fall_corr = .false.    ! Rain fall speed following ifs 
```
Should this comment be updated to say `Precip fall speed` instead of `Rain fall speed`?

nusbaume reviewed (approved):
Looks good!  Just one possible comment fix.

gold2718 reviewed (approved):
There are still unresolved conversations but it looks okay to me.

cacraigucar commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2665,13 +2665,13 @@ Default: .false.
 <entry id="micro_mg_ifs_sed" type="logical" category="microphys"
        group="micro_mg_nl" valid_values="" >
-Fixed sedimentation fall speeds for snow, ice and no sedimentation of liquid
+If .false., use the default sedimentation fall speed calculation from MG2 for all species. If .true., use constant sedimentation fall speeds for snow and ice, and no sedimentation of liquid, to be consistant with how species descend in the IFS model.
```
If I as a user wanted to get more information on the IFS model, could there be a little more explanation here?  A google search of "IFS model" doesn't give useful information on the first page.

gold2718 commented:
@brian-eaton, @nusbaume,
If issues have been resolved, please resolve the relevant conversations.
